README.md

```diff
@@ -1,38 +1,89 @@
+
+The exact components available to the extension depend on the surface area you are embedding — please refer to the [documentation linked above](#i-just-want-to-build-an-argo-extension-not-learn-about-them) for references on what components you can use. If you’ve ever used [Polaris](https://polaris.shopify.com/), though, the components you will find will look extremely familiar, as many components and props are based on their Polaris equivalents.
+
+### Extension points
+
+While remote-ui provides the component model, and takes care of propagating updates to an extension’s UI into the host application, it does not have any built-in notion of an “extension” at all. To create an extension system, Shopify authors a bit of extra code: a mapping of “extension points” (which, in code, are just strings with a specific naming format), and a way to load third party code that can register to be called for those extension points. This might sound little complicated, but the code to put it all together is actually quite straightforward. The Argo ”runtime” code, which is used to construct the sandbox your extension will run in, looks almost exactly like this for all Argo-capable surface areas in Shopify:
+
+```js
+// We keep a mapping of all the extension points you register for
+const registeredExtensionPoints = new Map();
+
+// We define a globally-available `shopify` object. The only thing this object
+// can do is register an extension (which is just a callback function) for a
+// specific extension point.
+globalThis.shopify = {
+  extend(extensionPoint, extension) {
+    registeredExtensionPoints.set(extensionPoint, extension);
+  },
+};
+
+// Here’s the function the host application will call to load your extension’s
+// script into the sandbox
+export function load(script) {
+  // Internal workings — don’t worry too much about this, because we might change
+  // it in the future as new web platform features are available!
+}
+
+// Once your script is loaded, it has had its chance to register for extension
+// points, and we can now simply call your extension with the arguments from the
+// host application, like the data your extension has access to and the UI “root“
+// your components will be attached to.
+export function run(extensionPoint, ...args) {
+  return registeredExtensionPoints.get(extensionPoint)?.(...args);
+}
+```
+
+In your extension code, you make use of these APIs using patterns that feel idiomatic in JavaScript:
+
+```js
+shopify.extend('Checkout::Feature::Render', (root) => {
+  root.appendChild(root.createText('Hello world!'));
+});
+
+// Or, if you are using our NPM libraries, we offer a bit of sugar over the
+// global:
+
+import {extend} from '@shopify/argo-checkout';
+
+extend('Checkout::Feature::Render', (root) => {
+  root.appendChild(root.createText('Hello world!'));
+});
+```
```

Comment:
It might be more helpful to put this code first as it illustrates the interface. The implementation model would make more sense following that.

Reply:
Hmm, I think that flow is kind of confusing, because we go from talking about the implementation in previous sections, to talking about extension code, to talking about the implementation again. I intended for the implementation of Argo to be the focus here; there's plenty of discussion that entirely focuses on the consumer code in the actual library documentation.